### PR TITLE
Optimize strain dependent rheology

### PR DESCRIPTION
**Pull request by gassmoeller**
This is an optimization of #4336. It makes the following changes:
- Using `FEPointEvaluation` instead of `FEFieldFunction` is much faster
- Only compute the solution values of one field instead of all fields, only at the current location, not at all in.positions
- Do not allocate std::vectors for every evaluation point, which are slow to create

@naliboff: Can you check that I didnt mess up the viscous vs plastic indices? I think I kept all functionality the same, it should just be significantly faster.

### CODE REVIEW

**Comment by naliboff** [comment]
@gassmoeller - Thanks! The indexing for the plastic and viscous strain case looks correct. However, some of the tests results are changing, but after a look at the changes most of the differences appear small.

 Can the use of `FEPointEvaluation `instead of `FEFieldFunction` produce small differences in the mapping of the old solution to the current positions, or should it give the same results?

**Comment by gassmoeller** [comment]
Thanks @naliboff, the test failures showed an actual bug (I had used only the plastic strain for the total strain healing option). This is fixed now and if all tests pass this should be ready for another review or merge.

File: source/material_model/rheology/strain_dependent.cc [hunk 1]
```diff
@@ -556,75 +556,113 @@ namespace aspect
...
-            Functions::FEFieldFunction<dim, LinearAlgebra::BlockVector>
-            fe_value(this->get_dof_handler(), this->get_old_solution(), this->get_mapping());
-            fe_value.set_active_cell(in.current_cell);
+            std::vector<Point<dim>> quadrature_positions(1,this->get_mapping().transform_real_to_unit_cell(in.current_cell, in.position[i]));
+            boost::container::small_vector<double, 100> old_solution_values(this->get_fe().dofs_per_cell);
+            in.current_cell->get_dof_values(this->get_old_solution(),
+                                            old_solution_values.begin(),
+                                            old_solution_values.end());
+
+            if (composition_evaluators.size() == 0)
```
Comment (naliboff) [inline source/material_model/rheology/strain_dependent.cc:567]: Question -  When would this condition be met? By default, isn't `composition_evaluators` of size `<1,dim>`?

File: source/material_model/rheology/strain_dependent.cc [hunk 2]
```diff
@@ -556,75 +556,113 @@ namespace aspect
...
-            std::vector<double> old_total_strain(in.n_evaluation_points());
-            std::vector<double> old_noninitial_plastic_strain(in.n_evaluation_points());
-
+            // from the old timesteps.
             // Prepare the field function and extract the old solution values at the current cell.
-            Functions::FEFieldFunction<dim, LinearAlgebra::BlockVector>
-            fe_value(this->get_dof_handler(), this->get_old_solution(), this->get_mapping());
-            fe_value.set_active_cell(in.current_cell);
+            std::vector<Point<dim>> quadrature_positions(1,this->get_mapping().transform_real_to_unit_cell(in.current_cell, in.position[i]));
+            boost::container::small_vector<double, 100> old_solution_values(this->get_fe().dofs_per_cell);
```
Comment (naliboff) [inline source/material_model/rheology/strain_dependent.cc:562]: Naive question - what is the eventual size of this vector? Is it `<1, dofs_per_cell>` or `<100,dofs_per_cell>`, or other? Is `100` just an arbitrary value to initialize vector, which is likely to never be exceeded?

File: source/material_model/rheology/strain_dependent.cc [hunk 3]
```diff
@@ -556,75 +556,113 @@ namespace aspect
...
+            in.current_cell->get_dof_values(this->get_old_solution(),
+                                            old_solution_values.begin(),
+                                            old_solution_values.end());
+
+            if (composition_evaluators.size() == 0)
+              composition_evaluators.resize(this->n_compositional_fields());
+
+            Assert(composition_evaluators.size() == this->n_compositional_fields(),
```
Comment (naliboff) [inline source/material_model/rheology/strain_dependent.cc:570]: I'm a bit confused by this Assert. composition_evaluators is created with 
`mutable std::vector<std::unique_ptr<FEPointEvaluation<1, dim>>> composition_evaluators;`. 

In this instance, would the above if statement be triggered? - ( `if (composition_evaluators.size() == 0)`). If not, when composition_evaluators be of size 0 and the resize line (`composition_evaluators.resize(this->n_compositional_fields());`) be triggered?

In summary, I'm struggling a bit to understand when and why this variable should be adjusted. A bit more documentation may help and apologies if I am just overlooking something obvious.

**Review by naliboff — COMMENTED** [review]
@gassmoeller - Thanks for the updates and finding the bug affecting the test results. I only have a few remaining questions regarding documentation, mainly to be sure I understand the logic if this section of code needs to be modified later on.

**Comment by gassmoeller** [comment]
I have added comments into the code to answer most of your questions. Let me know in case something is still unclear.

> Naive question - what is the eventual size of this vector? Is it <1, dofs_per_cell> or <100,dofs_per_cell>, or other? Is 100 just an arbitrary value to initialize vector, which is likely to never be exceeded?

boost::small_vector works the following way: It reserves a certain number of elements of the vector (in this case 100) on the stack, which is extremely fast. If the actual number of elements created (`dofs_per_cell`) is smaller than 100, it can purely use the stack to operate and requires no memory allocation. If `dofs_per_cell` is larger than 100, it needs to allocate elements on the heap (just like a normal `std::vector` always would), so it would be about as fast as a normal vector. But it will always have `dofs_per_cell` elements. In essence it is just a version of `std::vector` that is much faster to create and use, as long as it has less elements than a fixed number.

**Review by naliboff — APPROVED** [review]
@gassmoeller - Thanks for the additional documentation!
